// wa-sticker-formatter/dist/Sticker.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createSticker = exports.Sticker = void 0;
const fs_1 = require("fs");
const promises_1 = require("fs/promises");
const axios_1 = __importDefault(require("axios"));
const Utils_1 = __importDefault(require("./Utils"));
const file_type_1 = require("file-type");
const convert_1 = __importDefault(require("./lib/convert"));
const Exif_1 = __importDefault(require("./lib/Metadata/Exif"));
const StickerTypes_1 = require("./lib/Metadata/StickerTypes");
class Sticker {
    constructor(data, metadata = {}) {
        this.data = data;
        this.metadata = metadata;
        this._parse = () => __awaiter(this, void 0, void 0, function* () {
            return Buffer.isBuffer(this.data)
                ? this.data
                : yield (() => __awaiter(this, void 0, void 0, function* () {
                    return fs_1.existsSync(this.data)
                        ? promises_1.readFile(this.data)
                        : axios_1.default.get(this.data, { responseType: 'arraybuffer' }).then(({ data }) => data);
                }))();
        });
        this._getMimeType = (data) => __awaiter(this, void 0, void 0, function* () {
            const type = yield file_type_1.fromBuffer(data);
            if (!type)
                throw new Error('Invalid Buffer Instance');
            return type.mime;
        });
        /**
         * Builds the sticker
         * @param {string} [type] - How you want your sticker to look like
         * @returns {Promise<Buffer>} A promise that resolves to the sticker buffer
         */
        this.build = (type = this.metadata.type || StickerTypes_1.StickerTypes.DEFAULT) => __awaiter(this, void 0, void 0, function* () {
            const buffer = yield this._parse();
            const mime = yield this._getMimeType(buffer);
            return yield new Exif_1.default(this.metadata).add(yield convert_1.default(buffer, mime, type, this.metadata.quality));
        });
        /**
         * Saves the sticker to a file
         * @param [filename] - Filename to save the sticker to
         * @returns filename
         */
        this.toFile = (filename = this.defaultFilename) => __awaiter(this, void 0, void 0, function* () {
            yield promises_1.writeFile(filename, yield this.build());
            return filename;
        });
        /**
         * @deprecated
         * Use the `Sticker.build()` method instead
         */
        this.get = this.build;
        this.metadata.author = this.metadata.author || '';
        this.metadata.pack = this.metadata.pack || '';
        this.metadata.id = this.metadata.id || Utils_1.default.generateStickerID();
        this.metadata.quality = this.metadata.quality || 100;
        this.metadata.type = Object.values(StickerTypes_1.StickerTypes).includes(this.metadata.type)
            ? this.metadata.type
            : StickerTypes_1.StickerTypes.DEFAULT;
    }
    get defaultFilename() {
        return `./${this.metadata.pack}-${this.metadata.author}.webp`;
    }
}
exports.Sticker = Sticker;
const createSticker = (...args) => __awaiter(void 0, void 0, void 0, function* () {
    return yield new Sticker(...args).build();
});
exports.createSticker = createSticker;
